src/tree.rs

```diff
@@ -126,12 +129,161 @@ pub fn minimum_spanning_tree(
+fn _bipartition_tree(
+    spanning_tree: &graph::PyGraph,
+    pops: Vec<f64>,
+    pop_target: f64,
+    epsilon: f64,
+) -> Vec<(usize, Vec<usize>)> {
+    let mut pops = pops;
+    let spanning_tree_graph = &spanning_tree.graph;
+    let mut same_partition_tracker: Vec<Vec<usize>> =
+        vec![Vec::new(); spanning_tree_graph.node_bound()]; // Keeps track of all all the nodes on the same side of the partition
+
+    let mut node_queue: VecDeque<NodeIndex> = VecDeque::<NodeIndex>::new();
+    for leaf_node in spanning_tree_graph.node_indices() {
+        if spanning_tree_graph.neighbors(leaf_node).count() == 1 {
+            node_queue.push_back(leaf_node);
+        }
+        same_partition_tracker[leaf_node.index()].push(leaf_node.index());
+    }
+
+    // BFS search for balanced nodes
+    let mut balanced_nodes: Vec<(usize, Vec<usize>)> = vec![];
+    let mut seen_nodes = HashSet::with_capacity(spanning_tree_graph.node_count());
+    while !node_queue.is_empty() {
+        let node = node_queue.pop_front().unwrap();
+        if seen_nodes.contains(&node.index()) {
+            continue;
+        }
+
+        // Mark as seen; push to queue if only one unseen neighbor
+        let unseen_neighbors: Vec<NodeIndex> = spanning_tree
+            .graph
+            .neighbors(node)
+            .filter(|node| !seen_nodes.contains(&node.index()))
+            .collect();
+
+        if unseen_neighbors.len() == 1 {
+            // At leaf, will be false at root
+            let pop = pops[node.index()];
+
+            // Update neighbor pop
+            let neighbor = unseen_neighbors[0];
+            pops[neighbor.index()] += pop;
+
+            // Check if balanced; mark as seen
+            if pop >= pop_target * (1.0 - epsilon) && pop <= pop_target * (1.0 + epsilon) {
+                balanced_nodes.push((node.index(), same_partition_tracker[node.index()].clone()));
+            }
+            seen_nodes.insert(node.index());
+
+            // Update neighbor partition tracker
+            let mut current_partition_tracker =
+                mem::take(&mut same_partition_tracker[node.index()]);
+            same_partition_tracker[neighbor.index()].append(&mut current_partition_tracker);
+
+            // Queue neighbor
+            node_queue.push_back(neighbor);
+        } else if unseen_neighbors.is_empty() {
+            // Is root
+            break;
+        } else {
+            // Not a leaf yet
+            continue;
+        }
+    }
+
+    balanced_nodes
+}
+
+/// Bipartition graph into two contiguous, population-balanced components using
+/// mst. Assumes that the graph is contiguous. See :func:`~bipartition_tree` for
+/// details on how balance is defined.
+///
+/// :param PyGraph graph: Undirected graph
+/// :param weight_fn: A callable object (function, lambda, etc) which
+///     will be passed the edge object and expected to return a ``float``. See
+///     :func:`~minimum_spanning_tree` for details.
```

Comment:
This might be me bing picky, but I'd rename to something along the lines of `weight_sample_fn` and describe it as you described to me in the previous comment. It would make it clearer to the users and maintainers that the intended is non deterministic.

I'm saying so because in general, we expect `weight_fn` to always return the same result given the same edge. Moreover, wherever possible we also try to cache the calls to `weight_fn` to avoid calling Python from Rust more than necessary.